### PR TITLE
Add initial support for Remarkable Paper Pro

### PR DESCRIPTION
**Pull request by alexander-akhmetov**
A lot of it is thanks to the incredible work of many contributors from https://github.com/owulveryck/goMarkableStream/issues/117, including @owulveryck, @cristiannobili, @danielealbano, @To-Si, @Jayy001, @hegzploit and more 😄 

#### What currently does not work

1. RLE is not supported
2. The `Rotate` button in the UI
3. The `Colors` button

At least this, maybe something else. 

### Main changes

Because RLE isn't supported, I added optional zstd compression middleware alongside the existing gzip. It’s faster and should use less cpu.

#### ARM64 support

- Added `const_arm64.go` with device-specific constants for the Remarkable Paper Pro.
- `fb_arm.go` renamed to `fb_rm.go` and the build tags updated to support both arm and arm64.

#### Framebuffer pointer calculation for RMPP

The logic is taken from https://github.com/owulveryck/goMarkableStream/issues/117#issuecomment-2552499815 and https://github.com/owulveryck/goMarkableStream/issues/117#issuecomment-2497802044.

#### Client side (JS) changes

- Passing `screenWidth`, `screenHeight`, `UseRLE`, etc. from the backend to the JS client.
- Fixes for the texture coordinates and the color format if the device is RMPP.

#### Input devices

RMPP uses `/dev/input/event2` and `/dev/input/event3`, so added these to the device-specific constants.

---

To run the app on Paper Pro you need to disable the RLE and you can optionally enable the compression:

```
ZSTD_COMPRESSION=true RLE_COMPRESSION=false ./goMarkableStream
```

There's also a good possibility that the Remarkable 2 is broken. I'd appreciate it if someone could check.

### CODE REVIEW

**Comment by Jayy001** [comment]
This is some impressive work! Well done.

I've tried to run it on my own device, but unfortunately I'm having some issues. 

I ran `make build-remarkable-pro` which generated me the binary with md5 `1fe7dba1b23a936c244097be9744462a` and general info `goMarkableStream: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, Go, BuildID=pGPbavxPhEahbbmycnBO/KOuQ8CHlRzAuYsZNSJGr/7B5fNzvlIaUfr6NS1KT8/K2a6oZVCN8UYHf06abdl, stripped`. 

My device is on the beta version 17. When running, it displays the IP address but nothing else
```
root@imx8mm-ferrari:~# ZSTD_COMPRESSION=true RLE_COMPRESSION=false ./goMarkableStream
Local IP address: 10.11.99.1
Local IP address: 192.168.0.244
```
No service is running on port `2001`. 

Are there any debugging options to generate some useful output to help rectify this issue?

**Comment by hegzploit** [comment]
> This is some impressive work! Well done.
> 
> I've tried to run it on my own device, but unfortunately I'm having some issues.
> 
> I ran `make build-remarkable-pro` which generated me the binary with md5 `1fe7dba1b23a936c244097be9744462a` and general info `goMarkableStream: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, Go, BuildID=pGPbavxPhEahbbmycnBO/KOuQ8CHlRzAuYsZNSJGr/7B5fNzvlIaUfr6NS1KT8/K2a6oZVCN8UYHf06abdl, stripped`.
> 
> My device is on the beta version 17. When running, it displays the IP address but nothing else
> 
> ```
> root@imx8mm-ferrari:~# ZSTD_COMPRESSION=true RLE_COMPRESSION=false ./goMarkableStream
> Local IP address: 10.11.99.1
> Local IP address: 192.168.0.244
> ```
> 
> No service is running on port `2001`.
> 
> Are there any debugging options to generate some useful output to help rectify this issue?

Works fine on my device which is running 3.16.2.3.

**Comment by Jayy001** [comment]
That's good, what output does it give?

**Comment by Jayy001** [comment]
I've just downgraded my device to 3.16 and I can confirm it works fine, looks like potentially something on the beta is breaking it. If anybody else is on v17, would appreciate the test.
```
root@imx8mm-ferrari:~# ZSTD_COMPRESSION=true RLE_COMPRESSION=false ./goMarkableStream
Local IP address: 192.168.0.244
Local IP address: 10.11.99.1
2024/12/21 16:20:05 listening on [::]:2001
```
Here's the output it should give.

**Comment by Jayy001** [comment]
Animation
The output is...quite slanted, though.

**Comment by alexander-akhmetov** [comment]
That's me, in the last commit I broke it by changing the screen size constants. Can you check again please? I pushed the [fix](https://github.com/owulveryck/goMarkableStream/pull/122/commits/f7afec663fcb8cd4626e520969cdfc638ea10569).

**Comment by Jayy001** [comment]
Whooo, that worked!

**Comment by owulveryck** [comment]
Awesome;
I will test on my RM2 by next week, and it it works I will merge and release a new version.
Many thanks.

File: client/main.js [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
-const height = 1404;
-
-const rawCanvas = new OffscreenCanvas(width, height); // Define width and height as needed
+const rawCanvas = new OffscreenCanvas(screenWidth, screenHeight); // Define width and height as needed
 let portrait = getQueryParam('portrait');
 portrait = portrait !== null ? portrait === 'true' : false;
-let flip = getQueryParam('flip');
-flip = flip !== null ? flip === 'true' : false;
+
+defaultFlip = true;
```
Comment (owulveryck) [inline client/main.js:5]: On RM2 the defaultFlip should be false as well (otherwise the image is upside down)

**Review by owulveryck — APPROVED** [review]
Very good work!

File: client/main.js (hunk 1 above)
Comment (owulveryck) [reply client/main.js:5]: I will change it

**Comment by ArthurHlt** [comment]
> I've just downgraded my device to 3.16.1.0 and I can confirm it spins up the webserver, looks like potentially something on the beta is breaking it. If anybody else is on v17, would appreciate the test.
> 
> ```
> root@imx8mm-ferrari:~# ZSTD_COMPRESSION=true RLE_COMPRESSION=false ./goMarkableStream
> Local IP address: 192.168.0.244
> Local IP address: 10.11.99.1
> 2024/12/21 16:20:05 listening on [::]:2001
> ```
> 
> Here's the output it should give.

I've tested also on v17 and it is blocked after giving local ip address (but this information is very early in the code). So, for now i don't know what is blocking

**Pull request merged**